cub/test/catch2_test_device_reduce_deterministic.cu

```diff
@@ -0,0 +1,276 @@
+/******************************************************************************
+ * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#include "insert_nested_NVTX_range_guard.h"
+// above header needs to be included first
+
+#include <cub/detail/rfa.cuh>
+
+#include <thrust/device_vector.h>
+
+#include <algorithm>
+#include <numeric>
+
+#include "c2h/vector.cuh"
+#include "catch2_test_helper.h"
+#include "catch2_test_launch_helper.h"
+#include "cub/block/block_load.cuh"
+#include "cub/device/dispatch/dispatch_reduce_deterministic.cuh"
+#include "cub/thread/thread_load.cuh"
+#include <catch2/catch.hpp>
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+using float_type_list = c2h::type_list<float, double>;
+
+template <typename InputIteratorT, typename OutputIteratorT, typename NumItemsT>
+CUB_RUNTIME_FUNCTION static cudaError_t DeterministicSum(
```

Comment:
This might sound naive but I am a bit confused. Why are we declaring `DeterministicSum` in a test file? Don't we want this to be some interface under a header? If that's how the users are supposed to be using it the call to `cub::detail::DeterministicDispatchReduce::Dispatch` still gives me the creeps.

Reply:
@gevtushenko and others have not yet finalized the API and hence we have it here.

Reply:
The goal was to have minimal diff and no changes in existing codebase for now